### ISSUE TITLE
[re-S&C] [Design] [3/3] Prescribe fixes and solutions for Soil and Crop Module

### ISSUE THREAD

**Issue by PooyaHekmati** [RuminantFarmSystems/RuFaS#1236]
This is the third and last action item for _Analyzing S&C Module and planning on action items_ in the roadmap. Here, the goal is to come up with action items to improve the soil and crop module. This is done by comparing the results of #1230 and #1233 and the code and any other documentation. The emphasis is more on answering _what should be done_ versus _how we do it._
Put the created files in the corresponding folder in https://drive.google.com/drive/folders/1veAqYqIpXUPkOlTb7umz1wat0fd6THGQ?usp=sharing . 
Since there won't be any PRs for this, ask for a review before closing the issue.

**Comment by PooyaHekmati** [comment]
Blocked by #1230 and #1233 .

**Comment by allisterakun** [comment]
Issues #1104 can be taken care of here.

**Comment by Isgotkowitz** [comment]
Consider:
- simplifying the structure of the code
- must meet the requirements
- do not be bound by existing code structure (think outside of the box), but try to reuse as much code as possible
- maintain separation of responsibility between modules
- regulate data flow within RuFaS
  - all components at the same level of abstraction must communicate through a component that is a level of abstraction higher
  - levels of abstraction must not be mixed
- improving performance (including memory usage, runtime, etc.)
- unify documentation format
- enforce type consistency

**Comment by Isgotkowitz** [comment]
This issue is actually in-progress. The remaining tasks are:
- Review and finalize the prescribed fixes.
- Write up issues for the prescribed fixes.

**Comment by Isgotkowitz** [comment]
@ew3361zh I think the prescribed fixes look good!

I think the only issue that will need a bit more attention is #1544 - this more or less addresses the points from the above comment: 

> * regulate data flow within RuFaS
>   * all components at the same level of abstraction must communicate through a component that is a level of abstraction higher
>   * levels of abstraction must not be mixed

The issue has a bunch of sub-tasks, but these tasks have not been turned into their own issues yet. They are sort of tied to the issues related to breaking up the mega classes and dataclasses (#1815, #1816, #1820, #1821, #1822) but their scope extends further.

TL;DR the submodules within Crop and Soil are entangled, we want them disentangled similar to the biophysical modules (#1622).

**Comment by ew3361zh** [comment]
Added some issues to cover disentanglement of the above dataclasses a bit more. Not sure about needing further disentanglement with the Tillage and Fertilizer classes referenced in #1544 but we can discuss more in grooming sessions.
